core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -0,0 +1,251 @@
+
+    /** The last read-position in the log file */
+    private var logFileOffset = 0L
+
+    protected val logsRetryCount = 15
+    protected val logsRetryWait = 100.millis
+
+    /** HTTP connection to the container, will be lazily established by callContainer */
+    private var httpConnection: Option[HttpUtils] = None
+
+    def halt()(implicit transid: TransactionId): Future[Unit] = runc.pause(id)
+    def resume()(implicit transid: TransactionId): Future[Unit] = runc.resume(id)
+    def destroy()(implicit transid: TransactionId): Future[Unit] = docker.rm(id)
+
+    def initialize(initializer: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[Interval] = {
+        val start = transid.started(this, LoggingMarkers.INVOKER_ACTIVATION_INIT, s"sending initialization to $id $ip")
+
+        val body = JsObject("value" -> initializer)
+        callContainer("/init", body, timeout, retry = true).andThen { // never fails
+            case Success(r: RunResult) =>
+                transid.finished(this, start.copy(start = r.interval.start), s"initialization result: ${r.toBriefString}", endTime = r.interval.end)
+            case Failure(t) =>
+                transid.failed(this, start, s"initializiation failed with $t")
+        }.flatMap { result =>
+            if (result.ok) {
+                Future.successful(result.interval)
+            } else if (result.interval.duration >= timeout) {
+                Future.failed(InitializationError(ActivationResponse.applicationError(Messages.timedoutActivation(timeout, true)), result.interval))
+            } else {
+                Future.failed(InitializationError(ActivationResponse.processInitResponseContent(result.response, log), result.interval))
+            }
+        }
+    }
+
+    def run(parameters: JsObject, environment: JsObject, timeout: FiniteDuration)(implicit transid: TransactionId): Future[(Interval, ActivationResponse)] = {
+        val actionName = environment.fields.get("action_name").map(_.convertTo[String]).getOrElse("")
+        val start = transid.started(this, LoggingMarkers.INVOKER_ACTIVATION_RUN, s"sending arguments to $actionName at $id $ip")
+
+        val parameterWrapper = JsObject("value" -> parameters)
+        val body = JsObject(parameterWrapper.fields ++ environment.fields)
+        callContainer("/run", body, timeout, retry = false).andThen { // never fails
+            case Success(r: RunResult) =>
+                transid.finished(this, start.copy(start = r.interval.start), s"running result: ${r.toBriefString}", endTime = r.interval.end)
+            case Failure(t) =>
+                transid.failed(this, start, s"initializiation failed with $t")
+        }.map { result =>
+            val response = if (result.interval.duration >= timeout) {
+                ActivationResponse.applicationError(Messages.timedoutActivation(timeout, false))
```

Comment:
this is a corner case here where duration = timeout and result.response is valid.